### PR TITLE
Event prop

### LINKED ISSUE #56: CompiledTrainingNetwork train needs to support validation
- Keras allows either``validation_split`` and ``validation_data`` to be specified
- Compilers probably need to specify different lists of callbacks in each case (what did @tnowotny do in EventProp code base?)

### PR DESCRIPTION
**Pull request by neworderofjamie**
Aside from the EventProp compiler itself this adds:
- Some new readouts (readouts provided a real-valued output from a neuron) - ``AvgVar``, ``MaxVar`` and ``AvgVarExpWeight`` to reflect choices EventProp offers
- Previous softmax system was very e-prop centric. Moved guts of it into ``add_softmax_custom_updates`` helper and call this when softmax is required
- Switch ``SparseCategoricalAccuracy`` metric to use ``VarAccess_READ_ONLY_SHARED_NEURON`` variables rather than an EGP (allowing them to be double-buffered using a custom update)
- Added ``utils.data.linear_latency_encode_data`` as an alternative means of spike-encoding datasets
- Added kwargs to ``Readout.add_readout_logic`` to enable compilers to customise readouts e.g. to provide an example time for ``AvgVar`` and to set whether ``MaxVar`` records time at which maximum occurs
- EventProp examples for SHD and latency MNIST (also moved tonic data directory up into examples so all examples can share it)
- Support for generating a validation split when training or using a whole separate dataset

Fixes #56

### CODE REVIEW

File: ml_genn/ml_genn/compilers/compiler.py [hunk 1]
```diff
@@ -80,28 +82,71 @@ def create_reset_custom_update(reset_vars, var_ref_creator):
...
+        # the highest priority from possible
+        else:
+            return min(possible, key=lambda p: self._supported[p])
+        
+class ZeroInSyn(Callback):
+    def __init__(self, genn_syn_pop, example_timesteps: int):
+        self.genn_syn_pop = genn_syn_pop
+        self.example_timesteps = example_timesteps
+
+    def on_timestep_begin(self, timestep: int):
```
Comment (tnowotny) [inline ml_genn/ml_genn/compilers/compiler.py:132]: I am obviously naive, but does this mean in_syn is reset at the beginning of the last timestep of a trial? Not at the end?

File: ml_genn/ml_genn/readouts/avg_var.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        if self.output_var_name is None:
+            raise RuntimeError("AvgVar readout requires that models "
+                               "specify an output variable name")
+
+        # Find output variable
+        try:
+            output_var = [v for v in model.model["var_name_types"]
+                          if v[0] == self.output_var_name]
+        except StopIteration:
+            raise RuntimeError(f"Model does not variable "
```
Comment (tnowotny) [inline ml_genn/ml_genn/readouts/avg_var.py:25]: model does not *have* ...?

File: ml_genn/ml_genn/readouts/avg_var_exp_weight.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+        if self.output_var_name is None:
+            raise RuntimeError("AvgVarExpWeight readout requires that models "
+                               "specify an output variable name")
+
+        # Find output variable
+        try:
+            output_var = [v for v in model.model["var_name_types"]
+                          if v[0] == self.output_var_name]
+        except StopIteration:
+            raise RuntimeError(f"Model does not variable "
```
Comment (tnowotny) [inline ml_genn/ml_genn/readouts/avg_var_exp_weight.py:25]: *have*

File: ml_genn/ml_genn/readouts/max_var.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        if self.output_var_name is None:
+            raise RuntimeError("MaxVar readout requires that models "
+                               "specify an output variable name")
+
+        # Find output variable
+        try:
+            output_var = [v for v in model.model["var_name_types"]
+                          if v[0] == self.output_var_name]
+        except StopIteration:
+            raise RuntimeError(f"Model does not variable "
```
Comment (tnowotny) [inline ml_genn/ml_genn/readouts/max_var.py:25]: uhm ... some message about variable to take max of ...

File: ml_genn/ml_genn/compilers/event_prop_compiler.py [hunk 5]
```diff
@@ -0,0 +1,783 @@
...
+                        $(LambdaV) += (1.0 / $(RingIMinusV)[ringOffset + $(RingReadOffset)]) * ($(Vthresh) * $(LambdaV) + $(RevISyn));
+                        """
+
+                    # List of variables aside from those in base 
+                    # model we want to reset every batch
+                    additional_reset_vars = [("LambdaV", "scalar", 0.0),
+                                             ("LambdaI", "scalar", 0.0)]
+
+                    # If regularisation is enabled
+                    # **THINK** is this LIF-specific?
```
Comment (tnowotny) [inline ml_genn/ml_genn/compilers/event_prop_compiler.py:550]: I have never fully worked out the correct math on this but I think it is not LIF-specific.

File: ml_genn/ml_genn/compilers/event_prop_compiler.py [hunk 6]
```diff
@@ -0,0 +1,783 @@
...
+        $write
+        $$(RingWriteOffset)++;
+        
+        // Loop around if we've reached end of circular buffer
+        if ($$(RingWriteOffset) >= $max_spikes) {
+            $$(RingWriteOffset) = 0;
+        }
+    }
+    else {
+        //printf("%f: hidden: ImV buffer violation in neuron %d, fwd_start: %d, new_fwd_start: %d, rp_ImV: %d, wp_ImV: %d\\n", $$(t), $$(id), $$(RingReadEndOffset), $$(RingWriteStartOffset), $$(RingReadOffset), $$(RingWriteOffset));
```
Comment (tnowotny) [inline ml_genn/ml_genn/compilers/event_prop_compiler.py:272]: I see you left this here - what is your opinion about offering some sort of "strict buffer checking" flag that would enable this code?

**Review by tnowotny — APPROVED** [review]
Ok, I have read this one in some detail and while I wouldn't be able to pick up typos or involved errors I was able to follow the overall design and was able to recognise the eventprop elements. I have left some comments about minor details. 
I am not sure how I feel about leaving out avg_crossentropy as a potential loss and the other regularisation schemes. It's ok for now but we may want to revisit this question when we use mlgenn in anger. I will click approve though I note that atm the tests seem to fail.

File: ml_genn/ml_genn/compilers/compiler.py (hunk 1 above)
Comment (neworderofjamie) [reply ml_genn/ml_genn/compilers/compiler.py:132]: That was my attempt to duplicate https://github.com/tnowotny/genn_eventprop/blob/main/simulator_SHD.py#L1238-L1244

File: ml_genn/ml_genn/compilers/event_prop_compiler.py (hunk 6 above)
Comment (neworderofjamie) [reply ml_genn/ml_genn/compilers/event_prop_compiler.py]: Good idea - have added a ``strict_buffer_checking`` flag

File: examples/event_prop/shd.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+from time import perf_counter
+from ml_genn.utils.data import (calc_latest_spike_time, calc_max_spikes,
+                                preprocess_tonic_spikes)
+
+NUM_HIDDEN = 256
+BATCH_SIZE = 32
+NUM_EPOCHS = 300
+EXAMPLE_TIME = 20.0
+DT = 1.0
+SPARSITY = 1.0
```
[the remark this answers is not in the document]
Comment (neworderofjamie) [reply examples/event_prop/shd.py]: Done

File: ml_genn/ml_genn/readouts/avg_var.py (hunk 2 above)
Comment (neworderofjamie) [reply ml_genn/ml_genn/readouts/avg_var.py]: Good spot - all these error messages were badly mangled and the actual code to find the output variable also didn't work - that code will never raise ``StopIteration``

**Comment by neworderofjamie** [comment]
I am going to leave other regularisation schemes for now but I have reverted my removal of 'per_timestep_loss' into a new branch which I'll PR once this one is merged

File: ml_genn/ml_genn/compilers/compiler.py (hunk 1 above)
Comment (tnowotny) [reply ml_genn/ml_genn/compilers/compiler.py:132]: oh, yes, I had completely forgotten about that. Would it be worth copying the comment about clamping on last time step and  not bleeding spikes into the next trial?

File: ml_genn/ml_genn/compilers/event_prop_compiler.py (hunk 6 above)
Comment (tnowotny) [reply ml_genn/ml_genn/compilers/event_prop_compiler.py]: prefect!

File: ml_genn/ml_genn/compilers/compiler.py (hunk 1 above)
Comment (neworderofjamie) [reply ml_genn/ml_genn/compilers/compiler.py:132]: Yes, I had also clearly forgotten exactly what it was doing when I moved it out of EventProp - it's quite a specific thing so will move it back, rename and add a comment

Comment (neworderofjamie) [reply ml_genn/ml_genn/compilers/compiler.py]: You clearly know the right loose threads to tug on! The other use of ``ZeroInSyn`` in ``InferenceCompiler`` was also flawed - now fixed
